File: test/perf.jl
using ROOT
ROOT.gROOT.process_line("gErrorIgnoreLevel = 3001;")

function test_hist1()
    hi = TH1D("my_hist", "My Hist", Int32(10), -30.0, 30.0)
    n=100000000
    @inbounds for i=1:n
    	Fill(hi, randn())
    end
end

function test_hist2()
    hm = Dict{Int64, TH1D}()
    for j=1:100
        hm[j] = TH1D("my_hist_$j", "My Hist", Int32(10*j), -30.0, 30.0)
    end
    hi = TH1D("my_hist", "My Hist", Int32(10), -30.0, 30.0)
    n=100000
    @inbounds for i=1:n
    	for j=1:100
            Fill(hm[j], rand())
        end
    end
end


function test_hist3()
    hi = TH1D("my_hist", "My Hist", Int32(10), -30.0, 30.0)
    n=10000000
    ih = 0.0
    @inbounds for i=1:n
    	Fill(hi, randn())
        ih += Integral(hi)
        ih += GetMean(hi)
        ih += GetRMS(hi)
    end
end


function test_ttree1()
    tf = TFile("test.root", "RECREATE")
    ttree = TTree("my_tree", "My Tree")

    #branch variable should be array with length 1
    p1 = Float64[0.0]
    p2 = Int32[0]
    p3 = Float32[0.0]

    Branch(ttree, "p1", convert(Ptr{Void}, pointer(p1)), "p1/D")
    Branch(ttree, "p2", convert(Ptr{Void}, pointer(p1)), "p2/I")
    Branch(ttree, "p3", convert(Ptr{Void}, pointer(p1)), "p3/F")

    sum_entries = 0.0
    for i=1:10000000
        p1[1] = i
        p2[1] = i*2
        p3[1] = i*3
        Fill(ttree)
    end
    Write(tf)
    Close(tf)
end

function test_ttree2()
    tf = TFile("test.root")
    ttree = root_cast(TTree, Get(tf, "my_tree"))

    #branch variable should be array with length 1
    p1 = Float64[0.0]
    p2 = Int32[0]
    p3 = Float32[0.0]

    SetBranchAddress(ttree, "p1", convert(Ptr{Void}, pointer(p1)))
    SetBranchAddress(ttree, "p2", convert(Ptr{Void}, pointer(p1)))
    SetBranchAddress(ttree, "p3", convert(Ptr{Void}, pointer(p1)))

    s = 0.0
    for i=1:GetEntries(ttree)
        GetEntry(ttree, i)
        s += p1[1] + p2[1] + p3[1]
    end
    Close(tf)
end

function best(f)
    f()
    vs = Float64[]
    for i=1:5
        push!(vs, (@timed f())[2])
    end
    return round(mean(vs), 4)
end

println("hist1: ", best(test_hist1))
println("hist2: ", best(test_hist2))
println("hist3: ", best(test_hist3))
println("ttree1: ", best(test_ttree1))
println("ttree2: ", best(test_ttree2))

File: test/perf.cc
#include "TH1D.h"
#include "TTree.h"
#include "TFile.h"
#include "TROOT.h"
#include <random>
#include <sys/time.h>
#include <iostream>
#include <vector>
#include <map>
#include <string>

double clock_now()
{
    struct timeval now;

    gettimeofday(&now, NULL);
    return (double)now.tv_sec + (double)now.tv_usec/1.0e6;
}

void test_hist1() {
    TH1D* hi = new TH1D("my_hist", "My Hist", 10, -30.0, 30.0);
    int n = 100000000;
    for(int i=0;i<n; i++) {
        hi->Fill(random());
    }
    delete hi;
}

void test_hist2() {
    std::map<int, TH1D*> hm;
    for (int j=0;j<100;j++) {
        TH1D* hi = new TH1D(std::to_string(j).c_str(), "My Hist", 10*j, -30.0, 30.0);
        hm[j] = hi;
    }
    int n = 100000;
    for(int i=0;i<n; i++) {
        for (int j=0;j<100;j++) {
            hm[j]->Fill(random());
        }
    }
}

void test_hist3() {
    TH1D* hi = new TH1D("my_hist", "My Hist", 10, -30.0, 30.0);
    int n = 10000000;
    double ih = 0.0;
    for(int i=0;i<n; i++) {
        hi->Fill(random());
        ih = hi->Integral();
        ih += hi->GetRMS();
        ih += hi->GetMean();
    }
    delete hi;
}

void test_ttree1() {
    TFile* tf = new TFile("test.root", "RECREATE");
    TTree* ttree = new TTree("my_tree", "My Tree");

    double p1[1] = {0.0};
    int p2[1] = {0};
    float p3[1] = {0.0};

    ttree->Branch("p1", p1, "p1/D");
    ttree->Branch("p2", p2, "p2/I");
    ttree->Branch("p3", p3, "p3/F");
    
    for (int i=0; i<10000000; i++) {
        p1[0] = i;
        p2[0] = i*2;
        p3[0] = i*3;
        ttree->Fill();
    }
    tf->Write();
    tf->Close();
}

void test_ttree2() {
    TFile* tf = new TFile("test.root");
    TTree* ttree = (TTree*)(tf->Get("my_tree"));

    double p1[1] = {0.0};
    int p2[1] = {0};
    float p3[1] = {0.0};

    ttree->SetBranchAddress("p1", &p1);
    ttree->SetBranchAddress("p2", &p2);
    ttree->SetBranchAddress("p3", &p3);
    
    double s = 0.0;
    for (int i=0; i < ttree->GetEntries(); i++) {
        ttree->GetEntry(i);
        s += p1[0] + p2[0] + p3[0];
    }
    tf->Close();
}

double best(void (*f)(void) ) {
    std::vector<double> vs;
    for (int i=0; i<5; i++) {
        double t0 = clock_now();
        (*f)();
        double t1 = clock_now();
        vs.push_back(t1-t0);
    }
    double m = 0;
    for (auto& v : vs) {
        m += v;
    }
    return m / vs.size();
}

int main(int argc, const char** argv) {
    gROOT->ProcessLine("gErrorIgnoreLevel = 3001;");
    std::cout << "hist: " << best(&test_hist1) << std::endl;
    std::cout << "hist2: " << best(&test_hist2) << std::endl;
    std::cout << "hist3: " << best(&test_hist3) << std::endl;
    std::cout << "ttree1: " << best(&test_ttree1) << std::endl;
    std::cout << "ttree2: " << best(&test_ttree2) << std::endl;
}
